### PR TITLE
Make nightly parallel

### PR DESCRIPTION
**Pull request by guykeller**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/19749

## Description
1. Changed content's nightly build process to run on several machines in parallel.
2. Added an option to run content's nightly build on your local machine.

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

### CODE REVIEW

**Comment by guykeller** [comment]
FYI @yaakovi @ShahafBenYakir

File: Tests/test_content.py [hunk 1]
```diff
@@ -534,109 +709,187 @@ def execute_testing(server, server_ip, server_version, server_numeric_version, i
...
-        print("Some tests have failed. Not destroying instances.")
+
+def get_unmockable_tests(tests_settings):
+    conf, secret_conf = load_conf_files(tests_settings.conf_path, tests_settings.secret_conf_path)
+    unmockable_integrations = conf['unmockable_integrations']
+    tests = conf['tests']
+    unmockable_tests = []
+    for test_record in tests:
+        test_name = test_record.get("playbookID")
+        integrations_used_in_test = get_tested_integrations(test_record)
```
Comment (ronykoz) [inline Tests/test_content.py]: maybe change the function name to be `get_used_integrations`

File: Tests/test_content.py [hunk 2]
```diff
@@ -534,109 +709,187 @@ def execute_testing(server, server_ip, server_version, server_numeric_version, i
...
+        unmockable_integrations_used = [integration_name for integration_name in integrations_used_in_test if
+                                        integration_name in unmockable_integrations]
+        if test_name and (not integrations_used_in_test or unmockable_integrations_used):
```
Comment (ronykoz) [inline Tests/test_content.py]: how test_name can be false?

File: Tests/test_content.py [hunk 3]
```diff
@@ -486,24 +599,85 @@ def run_test_scenario(t, proxy, default_test_timeout, skipped_tests_conf, nightl
...
+
+
+def get_instances_ips_and_names():
+    with open('./Tests/instance_ips.txt', 'r') as instance_file:
+        instance_ips = instance_file.readlines()
+        instance_ips = [line.strip('\n').split(":") for line in instance_ips]
+        return instance_ips
+
+
+def get_tests_records_from_names(tests_settings, tests_names_to_search):
```
Comment (ronykoz) [inline Tests/test_content.py]: not a good name

File: Tests/test_content.py [hunk 4]
```diff
@@ -486,24 +599,85 @@ def run_test_scenario(t, proxy, default_test_timeout, skipped_tests_conf, nightl
...
+                    return
+                else:
+                    failed_verify_login_message = "Failed verifying login (will retry). status: {}. " \
+                                                  "text: {}".format(res.status_code, res.text)
+                    prints_manager.add_print_job(failed_verify_login_message, print, thread_index)
+            except Exception as ex:
+                error_message = "Failed verifying server start via login: {}".format(ex)
+                prints_manager.add_print_job(error_message, print_error, thread_index)
+
+    raise Exception('Timeout waiting for demisto service to restart')
```
Comment (ronykoz) [inline Tests/test_content.py]: i think that @avidan-H removed this

File: Tests/test_content.py [hunk 5]
```diff
@@ -52,53 +58,153 @@ def options_handler():
...
+class TestsDataKeeper:
+
+    def __init__(self):
+        self.succeeded_playbooks = []
+        self.failed_playbooks = []
+        self.skipped_tests = []
+        self.skipped_integrations = []
+        self.rerecorded_tests = []
+        self.empty_files = []
+        self.unmockable_integrations = {}
```
Comment (ronykoz) [inline Tests/test_content.py]: you can use those directly in your code and then you won't need the add_tests_data function

File: Tests/test_dependencies.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        for integration_name in record_integrations:
+            if integration_name in integration_to_tests_mapping:
+                if record_playbook_name not in integration_to_tests_mapping[integration_name]:
+                    integration_to_tests_mapping[integration_name].append(record_playbook_name)
+            else:
+                integration_to_tests_mapping[integration_name] = [record_playbook_name]
+    return integration_to_tests_mapping
+
+
+def get_tested_integrations(test_playbook_record):
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: bad name, please see the comment in test_content about this

File: Tests/test_dependencies.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    return dependent_tests, independent_tests, all_tests
+
+
+def get_dependent_integrations_clusters_data(tests_file_path):
+    tests_graph = TestsGraph()
+    tests_graph.build_tests_graph_from_conf_json(tests_file_path)
+    return tests_graph.clusters
+
+
+def get_tests_allocation(number_of_instances, tests_file_path):
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: what allocation exactly

File: Tests/test_dependencies.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+def get_test_dependencies(tests_file_path):
+    dependent_integrations = get_integration_dependencies(tests_file_path)[0]
+
+    with open(tests_file_path, 'r') as myfile:
+        conf_json_string = myfile.read()
+    conf_json_obj = json.loads(conf_json_string)
+
+    dependent_tests = []
+    all_tests = []
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: could be dependent_tests + independent_tests in the end
or just set

File: Tests/test_dependencies.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+def get_tested_integrations(test_playbook_record):
+    tested_integrations = test_playbook_record.get("integrations", [])
+    if isinstance(tested_integrations, list):
+        return tested_integrations
+    else:
+        return [tested_integrations]
+
+
+def get_integration_dependencies(tests_file_path):
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: what deps exactly?

File: Tests/test_dependencies.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    def get_clusters(self):
+        clusters = []
+        for test_name in self.test_vertices:
+            test_vertex = self.test_vertices[test_name]
+            if not test_vertex.visited:
+                test_connected_component = test_vertex.get_connected_component([])
+                clusters.append(test_connected_component)
+        self.clusters = clusters
+
+    def build_tests_graph_from_conf_json(self, tests_file_path):
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: you should receive `dependent_tests` you already have it when you launch this func

File: Tests/test_dependencies.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+    def get_clusters(self):
+        clusters = []
+        for test_name in self.test_vertices:
+            test_vertex = self.test_vertices[test_name]
+            if not test_vertex.visited:
+                test_connected_component = test_vertex.get_connected_component([])
+                clusters.append(test_connected_component)
+        self.clusters = clusters
+
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: need to go over

File: Tests/test_dependencies.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    tests_graph.build_tests_graph_from_conf_json(tests_file_path)
+    return tests_graph.clusters
+
+
+def get_tests_allocation(number_of_instances, tests_file_path):
+    dependent_tests, independent_tests, all_tests = get_test_dependencies(tests_file_path)
+    dependent_tests_clusters = get_dependent_integrations_clusters_data(tests_file_path)
+    dependent_tests_clusters.sort(key=len, reverse=True)  # Sort the clusters from biggest to smallest
+    tests_allocation = []
+    number_of_tests_left = len(all_tests)
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: shouldn't it be minus independent_tests ?

File: Tests/test_dependencies.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+            for tests_cluster in dependent_tests_clusters:
+                current_allocation.extend(tests_cluster)
+            for test_name in independent_tests:
+                current_allocation.append(test_name)
+            tests_allocation.append(current_allocation)
+            break
+
+        if len(dependent_tests_clusters) > 0:
+            first_cluster = dependent_tests_clusters.pop(0)
+            first_cluster_size = len(first_cluster)
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: redundant

File: Tests/test_dependencies.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+        if current_allocation_size > desired_tests_per_allocation:
+            tests_allocation.append(current_allocation)
+            continue
+
+        clusters_added = 0
+        for cluster in dependent_tests_clusters:
+            cluster_size = len(cluster)
+            if current_allocation_size + cluster_size > desired_tests_per_allocation:
+                break
```
Comment (ronykoz) [inline Tests/test_dependencies.py]: explain this

File: Tests/test_dependencies.py (hunk 7 above)
Comment (ronykoz) [inline Tests/test_dependencies.py]: lets go over this together

File: Tests/test_content.py (hunk 4 above)
Comment (avidan-H) [reply Tests/test_content.py]: yea, I don't think restarting demisto service is necessary. It stam eats up time. Can make an API call to demisto client to reset containers instead.

File: Tests/test_dependencies.py (hunk 11 above)
Comment (guykeller) [reply Tests/test_dependencies.py]: Done

File: .circleci/config.yml [hunk 15]
```diff
@@ -18,35 +18,35 @@ jobs:
...
-            python3 --version
+              if [ -n "${INSTANCE_TESTS}" ] && [ $CIRCLE_NODE_INDEX -ne 0 ] ;
+                then
+                  echo "Skipping - instance tests tasks are running only in container number 0"
+                  exit 0
+              fi
+              echo 'export CIRCLE_ARTIFACTS="/home/circleci/project/artifacts"' >> $BASH_ENV
+              echo 'export PATH="/home/circleci/.local/bin:${PATH}"' >> $BASH_ENV # disable-secrets-detection
+              echo 'export PYTHONPATH="/home/circleci/project:${PYTHONPATH}"' >> $BASH_ENV
+              echo 'export NIGHTLY="true"' >> $BASH_ENV
```
[the remark this answers is not in the document]
Comment (guykeller) [reply .circleci/config.yml]: At the end of everything

File: Tests/mock_server.py [hunk 16]
```diff
@@ -289,7 +294,8 @@ def start(self, playbook_id, path=None, record=False):
         if not log_file_exists:
             self.stop()
             raise Exception("Proxy process took to long to go up.")
-        print('Proxy process up and running. Took {} seconds'.format(seconds_since_init))
+        proxy_up_message = 'Proxy process up and running. Took {} seconds'.format(seconds_since_init)
+        prints_manager.add_print_job(proxy_up_message, print, thread_index)
```
Comment (ShahafBenYakir) [inline Tests/mock_server.py]: In several functions headers, `prints_manager=None` is being used, however, you are never addressing this case where `prints_manage` is actually `None`. For example here, you have `prints_manager.add_print_job(proxy_up_message, print, thread_index)` which could obviously fail if `prints_manager` is `None. It feels to me that we either should address this case where `prints_manager=None` or don't have the default at all. Tell me what you think.
This comment is relevant for every function that has `prints_manager=None` as a parameter.

File: Tests/test_content.py [hunk 17]
```diff
@@ -34,8 +40,7 @@
 
 def options_handler():
     parser = argparse.ArgumentParser(description='Utility for batch action on incidents')
-    parser.add_argument('-u', '--user', help='The username for the login', required=True)
-    parser.add_argument('-p', '--password', help='The password for the login', required=True)
+    parser.add_argument('-k', '--apiKey', help='The Demisto API key for the server', required=True)
     parser.add_argument('-s', '--server', help='The server URL to connect to')
```
Comment (ShahafBenYakir) [inline Tests/test_content.py]: Let's remove this server param

**Comment by ShahafBenYakir** [comment]
I know it's a file you haven't worked on, but, can we remove the `SERVER_IP` and `SERVER_URL` constants from the the `tun_tests.sh` script (line 6,7) we are not using them anymore and it throws a warning for nothing.

File: Tests/lastest_server_build_scripts/run_tests.sh [hunk 18]
```diff
@@ -4,14 +4,13 @@ echo "start content tests"
...
 CONF_PATH="./Tests/conf.json"
-USERNAME=$(cat $SECRET_CONF_PATH | jq '.username')
+DEMISTO_API_KEY=$(cat $SECRET_CONF_PATH | jq '.temp_apikey')
 
-# remove quotes from username
-temp="${USERNAME%\"}"
+temp="${DEMISTO_API_KEY%\"}"
 temp="${temp#\"}"
-USERNAME=$temp
+DEMISTO_API_KEY=$temp
```
Comment (ronykoz) [inline Tests/lastest_server_build_scripts/run_tests.sh]: revert this file

File: Tests/scripts/wait_until_server_ready.py [hunk 19]
```diff
@@ -33,20 +36,19 @@ def get_username_password():
         conf = json.load(conf_file)
 
     if options.non_ami:
-        return conf['username'], conf['username'], options.contentVersion
+        return conf['temp_apikey'], options.contentVersion
```
Comment (ronykoz) [inline Tests/scripts/wait_until_server_ready.py]: revert this line, server build will not have this api key

File: Tests/test_content.py (hunk 17 above)
Comment (ronykoz) [reply Tests/test_content.py]: i think that it is used for the non_ami ones

File: Tests/test_content.py [hunk 20]
```diff
@@ -50,55 +55,173 @@ def options_handler():
...
+        self.print_lock = threading.Lock()
+        self.threads_last_update_times = [time.time() for i in range(number_of_threads)]
+
+    def should_update_thread_status(self, thread_index):
+        current_time = time.time()
+        thread_last_update = self.threads_last_update_times[thread_index]
+        return current_time - thread_last_update > 300
+
+    def add_print_job(self, message_to_print, print_function_to_execute, thread_index, message_color=None):
+        if print_color:
```
Comment (ronykoz) [inline Tests/test_content.py]:
```suggestion
        if message_color:
```

File: Tests/test_content.py [hunk 21]
```diff
@@ -534,109 +708,203 @@ def execute_testing(server, server_ip, server_version, server_numeric_version, i
...
+                if ami_instance_name == server_version:  # Only run tests for server master
+                    current_instance = ami_instance_ip
+                    tests_allocation_for_instance = test_allocation[current_thread_index]
+
+                    unmockable_tests = [test for test in all_unmockable_tests_list if test in tests_allocation_for_instance]
+                    mockable_tests = [test for test in tests_allocation_for_instance if test not in unmockable_tests]
+                    print_color("Starting tests for {}".format(ami_instance_name), LOG_COLORS.GREEN)
+                    print("Starts tests with server url - https://{}".format(ami_instance_ip))
+
+                    if number_of_instances == 1:
```
Comment (ronykoz) [inline Tests/test_content.py]: shouldn't this be ?
```suggestion
                    if not is_nightly or number_of_instances == 1:
```
@yaakovi take a quick look here, i want to merge this

**Comment by ronykoz** [comment]
Force merging

**Pull request merged**